bots/run-queue

```diff
@@ -21,40 +21,72 @@ MAX_PRIORITY = 9
 
 import argparse
 import json
+import random
+import ssl
 import subprocess
 import sys
+import urllib.request, urllib.parse, urllib.error
+
+import pika
+from task import REDHAT_PING
+
+# Check if we have access to Red Hat network
+try:
+    urllib.request.urlopen(REDHAT_PING).read()
+    run_redhat_tasks = True
+except IOError:
+    run_redhat_tasks = False
 
-import amqp
 
 def main():
     parser = argparse.ArgumentParser(description='Bot: read a single test command from the queue and execute it')
-    parser.add_argument('--amqp', default='localhost',
-                        help='The URL of the AMQP server to consume from')
-    parser.add_argument('--queue', default='tasks',
-                        help='The name of the queue to consume from')
+    parser.add_argument('--amqp', default='localhost:5671',
+                        help='The host:port of the AMQP server to consume from (default: %(default)s)')
     opts = parser.parse_args()
 
-    connection = amqp.Connection(host=opts.amqp)
-    connection.connect()
+    try:
+        host, port = opts.amqp.split(':')
+    except ValueError:
+        parser.error('Please format --amqp as host:port\n')
+
+    connection = pika.BlockingConnection(pika.ConnectionParameters(
+        host=host,
+        port=int(port),
+        ssl=True,
+        ssl_options=pika.SSLOptions(
+            ssl_version=ssl.PROTOCOL_TLSv1_2,
+            cafile='/run/secrets/webhook/ca.pem',
+            keyfile='/run/secrets/webhook/amqp-client.key',
+            certfile='/run/secrets/webhook/amqp-client.pem',
+            server_hostname=host),
+        credentials=pika.credentials.ExternalCredentials()))
     channel = connection.channel()
+
     try:
-        channel.queue_declare(queue=opts.queue, passive=True, auto_delete=False)
-    except amqp.exceptions.NotFound:
-        sys.stdout.write('queue is empty\n')
+        declare_public_result = channel.queue_declare(queue='public', passive=True, auto_delete=False)
+        declare_rhel_result = channel.queue_declare(queue='rhel', passive=True, auto_delete=False)
+    except pika.exceptions.ChannelClosed:
+        print("One of the queues doesn't exist")
         return 0
 
-    channel.basic_qos(0, 1, True)
+    queue='public'
+    if run_redhat_tasks:
+        # Try the rhel queue if the public queue is empty
+        if declare_public_result.method.message_count == 0:
+            queue = 'rhel'
+        # If both are non-empty, shuffle
+        elif declare_rhel_result.method.message_count > 0:
+            queue = ['public', 'rhel'][random.randrange(2)]
 
     # Get one item from the queue if present
-    msg = channel.basic_get(queue=opts.queue)
-    if msg:
-        body = json.loads(msg.body)
+    method_frame, header_frame, body = channel.basic_get(queue=queue)
```

Comment:
So `basic_get()` returns a triple with `None` as first element if there is no queued item? As the logic above would let two empty queues pass.

Reply:
exactly, plus the queue might have changed between declaring and consuming and it may now be empty. The first check is really if the queue doesn't exist and passive=True so it won't create the queue.